src/category_theory/subobject.lean

```diff
@@ -0,0 +1,1330 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Scott Morrison
+-/
+import category_theory.opposites
+import category_theory.full_subcategory
+import category_theory.skeletal
+import category_theory.currying
+import category_theory.limits.lattice
+import category_theory.limits.over
+import category_theory.limits.shapes.images
+import category_theory.limits.shapes.kernels
+import category_theory.monad.adjunction
+
+/-!
+# The lattice of subobjects
+
+We define `subobject X` as the quotient (by isomorphisms) of
+`mono_over X := {f : over X // mono f.hom}`.
+
+Here `mono_over X` is a thin category (a pair of objects has at most one morphism between them),
+so we can think of it as a preorder. However as it is not skeletal, it is not a partial order.
+
+There is a coercion from `subobject X` back to the ambient category `C`
+(using choice to pick a representative), and for `P : subobject X`,
+`P.arrow : (P : C) ⟶ X` is the inclusion morphism.
+
+The predicate `h : P.factors f`, for `P : subobject Y` and `f : X ⟶ Y`
+asserts the existence of some `P.factor_thru f : X ⟶ (P : C)` making the obvious diagram commute.
+We provide conditions for `P.factors f`, when `P` is a kernel/equalizer/image/inf/sup subobject.
```

Comment:
```suggestion
We provide conditions for `P.factors f`, when `P` is a kernel/equalizer/image/inf/sup subobject.

TODO: Add conditions for when `P` is a pullback subobject.
```
(mostly because I'll certainly need this and I don't want to forget to do it)